[2단계 - 출석 다시 구현하기] 도기(이동엽) 미션 제출합니다.

dye0p opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [2단계 - 셀프체크](https://prolog.techcourse.co.kr/studylogs/3982)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
TDD에 익숙해지는 것이 목표였기 때문에 객체지향 생활 체조 원칙보단 TDD에 집중했습니다. TDD를 통해서 코드를 완성했을 때 설계의 허점을 발견했고, 객체지향적으로 잘 설계가 된것같지 않다는 느낌을 받았습니다. 메서드가 의미단위로 구별되지 못해서 메서드가 10줄을 넘기는 경우가 조금 많았습니다. 추가로 원시값 포장 등등 지키지 못한 부분이 많다고 생각합니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요 이프!  2단계 리뷰에 참여해주셔서 감사합니다🙇‍♂️ 
2단계를 진행하면서 고민했던 부분들을 적어보았습니다!

## 🤔 고민

### 설계와 TDD

TDD를 할때 설계를 최소한으로 가져가라는 것을 보고, 최소한의 설계를 하고 테스트 코드를 먼저 작성했습니다. 
제일 먼저 느꼈던건 `객체에게 적합한 책임이 있지 않다?` 였습니다. TDD를 진행할때 적합한 책임이라고 생각하며 테스트를 작성했지만 후에 보니 이후 컨트롤러(Controller)와 뷰(View)를 연결하는 과정에서 설계의 허점을 많이 발견할 수 있었습니다. 

그래서 TDD를 진행할때 설계의 범위를 어디까지 가져가야 하는지에 대한 의문이 들었습니다.

### DTO 제거

step1에서 도메인이 dto를 의존하는 문제를 도메인 외부에서 dto를 생성 해줄 수도 있었지만, 굳이 dto가 필요할까? 라는 생각이 들어 step2 에선 dto를 사용하지 않았습니다. 

dto를 사용하지 않아서 힘들었던 부분은 원하는 데이터를 운반할 수단이 없다는 것이 힘들게 다가왔습니다. 그래서 최대한 Map과 같은 컬렉션을 활용하여 데이터를 포장해줄 수 있도록 하였습니다. 

## 🖐️ 질문

1. 필요한 기능을 미리 예상하여 테스트 코드로 만드는 것 vs 프로덕션 코드를 작성하다 필요한 기능이 생기면 그때 테스트를 작성

> 2. TDD를 할때 최소한의 설계를 어디까지 가져가야 하는지?
> 예를들면 객체간의 협력까지 구상하고 가는지, 떠오르는 객체들만 구상하고 가는지
> 
> >  이 질문은 1단계에서 이미 답변을 주셨네요!

3. 현업에는 TDD를 어떻게 적용하여 사용하는지?

dye0p committed:
setting: step2를 위한 기존 파일 삭제

dye0p committed:
docs: 구현 기능 목록 추가

dye0p committed:
feat: 출석 저장 기능 추가

dye0p committed:
feat: 크루들의 출석 기록을 가지는 Attendances 객체 추가

dye0p committed:
feat: 닉네임과 일치하는 크루의 출석부를 가져오는 기능 추가

dye0p committed:
feat: 출석 등록 기능 추가

dye0p committed:
feat: 날짜(일)로 요일을 판단하는 기능 추가

- 공휴일인 경우 예외가 발생한다.
- 공휴일엔 주말(토,일), 지정공휴일(25일)이 포함된다.

dye0p committed:
test: Attendances checkAttendance메서드 테스트 반환타입 변경(AttendanceState -> void)

dye0p committed:
refactor: Calender 메서드 분리

dye0p committed:
feat: 출석 수정 기능 추가

dye0p committed:
feat: 출석 시간에 따른 출석 상태 계산 기능 추가

dye0p committed:
feat: 크루별 출석 기록 조회 기능 추가

dye0p committed:
feat: 제적 위험자 확인 기능 추가

dye0p committed:
feat: 누적 지각 및 결석 횟수로 경고 또는 면담 대상 계산 기능 추가

dye0p committed:
refactor: domain 객체 domain 패키지로 이동

dye0p committed:
refactor: Attendance 정적 팩토리 메서드 추가

dye0p committed:
refactor: Attendances findCrewBy 메서드명 변경 -> findAttendanceBy

dye0p committed:
feat: 파일 읽어오기 기능 추가

dye0p committed:
feat: 프로그램 실행을 위한 Application 클래스 추가

dye0p committed:
feat: 날짜 정보를 가지는 DateProvider 객체 추가

dye0p committed:
feat: 기능 입력 커맨드에 따른 기능실행 기능 추가

dye0p committed:
feat: 안내 메세지 출력과 기능 입력 기능 추가

dye0p committed:
feat: 캠퍼스 등교시간이 아닌 경우 예외처리 추가

dye0p committed:
refactor: Calender findBy() 메서드 오버로딩 중복 리팩터링

dye0p committed:
feat: 닉네임으로 크루를 못찾을시 예외처리 추가

dye0p committed:
feat: 출석 등록 입출력 기능 추가

dye0p committed:
feat: 출석 수정 입출력 기능 추가

dye0p committed:
feat: 출석 수정 닉네임 입력시 예외처리 기능 추가

dye0p committed:
feat: 전날까지의 크루 출석 기록 입출력 기능 추가

dye0p committed:
feat: 제적 위험자 결과 조회 입출력 기능 추가

dye0p committed:
docs: 예외처리 완료사항 docs 업데이트

dye0p committed:
refactor: DateProvider 클래스 네이밍 변경 후 Date 관련 책임 이동

- DateProvider -> AttendanceDate 로 변경

dye0p committed:
refactor: AttendanceController 메서드 분리 및 순서 정리

dye0p committed:
feat: 이미 출석한 크루에 대한 예외처리 추가

dye0p committed:
feat: 공휴일을 가지는 HolyDay 객체 추가

dye0p committed:
refactor: Attendance 메서드 순서 변경

dye0p committed:
refactor: OutPutView 메서드 분리 및 문자열 상수화

sinb57 commented on src/main/java/domain/Dismissal.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public static Dismissal findDismissalBy(final int lateCount, int absenceCount) {
+        absenceCount = calculateAbsenceCountBy(lateCount, absenceCount);
```
되도록 재할당하지 않는 습관을 갖는 것이 좋아요.

dye0p replied:
그렇네요! 매개변수의 값을 재할당하면 예상하지 못한 변경이 일어날 수 있겠네요! 피드백 반영하겠습니다 :)

sinb57 replied:
그런 의미에서 final 사용은 좋은 선택지예요! 재할당을 원초적으로 막아주거든요.
지금 final이 부분적으로 사용되고 있는데, 사용하고자 한다면 모든 곳에 사용하거나 아예 안하는 등 통일하는 것이 좋아요.
final이 선언되지 않은 변수는 오히려 재할당이 가능한 것이라고 착각할 수 있고, 이러한 작은 오해와 의심 하나하나가 생산성을 저하시킬 수 있어요.

sinb57 commented on src/main/java/domain/Crew.java:
```diff
@@ -7,10 +7,15 @@ public class Crew {
     }
 
-    public static Crew from(final String name) {
+    public static Crew of(String name) {
```
정적 팩토리 메서드를 만든 이유가 궁금해요~

dye0p replied:
생성자에서 아무일도 하지 않는다면 그냥 생성자를 사용해도 무방하겠지만 생성자에 이름을 부여하고 싶어서 사용했습니다!

sinb57 commented on src/main/java/domain/Attendances.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    public void checkAttendance(final String name, final LocalDateTime time) {
+        CampusTime.validateInTime(time.toLocalTime());
```
Attendances 도메인이 CampusTime에 의존해야 할까요? LocalDateTime 원시값을 포장하여 해당 책임을 위임하면 어떨까요?

dye0p replied:
LocalDate를 포장하는 AttendanceTime 객체를 사용하고 있는데 제대로 활용을 못 한것 같아요! 피드백 반영해보겠습니다!

sinb57 commented on src/test/java/domain/AttendanceStateTest.java:
```diff
@@ -3,33 +3,86 @@
+        assertThat(actual).isEqualTo(AttendanceState.LATE);
+    }
+
+    @DisplayName("날짜가 월요일인 경우 등교시간이 13:00라면 출석이다.")
+    @Test
+    void attendanceState() {
+        //given
+        LocalDateTime time = LocalDateTime.of(2024, 12, 2, 13, 0);
+
+        //when
+        AttendanceState attendanceState = AttendanceState.findStateBy(time);
+
+        //then
+        assertThat(attendanceState).isEqualTo(AttendanceState.ATTENDANCE);
+    }
+
+    @DisplayName("날짜가 월요일인 경우 등교시간이 13:05분을 초과하면 지각이다.")
+    @Test
+    void attendanceState2() {
+        //given
+        LocalDateTime time = LocalDateTime.of(2024, 12, 2, 13, 6);
+
+        //when
+        AttendanceState attendanceState = AttendanceState.findStateBy(time);
+
+        //then
+        assertThat(attendanceState).isEqualTo(AttendanceState.LATE);
```
테스트 구조가 반복되네요! parameterizedTest를 활용해보면 어떨까요? 필요하다면 Named를 사용해봐도 좋아요

dye0p replied:
Named는 한번도 사용하지 않았는데 한번 사용해보겠습니다!

sinb57 replied:
Nested를 잘 사용하셨지만, 의도한 바는 아니에요. 아래와 같이 파라미터를 제어해서 테스트를 구성하길 바랬어요.

```java
@DisplayName("날짜가 월요일인 경우 등교시간이 {}을 초과하면 {}이다.")
@ParameterizedTest
void attendanceState(LocalDateTime attendTime, AttendanceState expected) {
    AttendanceState attendanceState = AttendanceState.findStateBy(attendTime);

    assertThat(attendanceState).isEqualTo(expected);
}
```

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,147 @@
+
+    void attendanceUpdate() {
+        String name = inputUpdateName();
+        attendances.findAttendanceBy(name);
```
검증을 위한 메서드 호출인가요?

dye0p replied:
코드를 작성하고 보니 뭔가 어울리지 않는다는 느낌을 받았습니다. 메서드 네이밍과 예외 메세지 부분에서 매치가 안된다는 느낌을 받았고, 메서드 네이밍은 findAttendanceBy인데 내부 로직에서는 크루의 이름으로 값을 찾고 있는것 같네요ㅠ 피드백 반영하여 수정해보겠습니다!

sinb57 reviewed (changes requested):
안녕하세요 도기~ 연휴에도 2단계 진행하느라 고생 많으셨어요! 코멘트 확인 부탁드려요~
2단계 리뷰 요청 마감일이 지났기 때문에, 다음 미션에 집중할 수 있도록 다음 혹은 다다음에 머지할게요.
얘기 나누고 싶은 부분이 있다면 코멘트 남겨주세요~

---

> 1. 필요한 기능을 미리 예상하여 테스트 코드로 만드는 것 vs 프로덕션 코드를 작성하다 필요한 기능이 생기면 그때 테스트를 작성

말씀주신 2가지 방법은 함께 사용해야 하지 않나 싶어요. 테스트 코드를 작성한 후 프로덕션 코드를 구현하더라도, 구현 과정에서 필요한 기능이 추가로 발생할 수 있고 그럼 추가 기능에 대한 테스트 코드를 작성하게 될테니까요.

> 3. 현업에는 TDD를 어떻게 적용하여 사용하는지?

시간적 여유가 있다면 테스트코드와 더불어 구현을 진행하곤 하지만 일정상 테스트 작성할 시간이 없는 경우도 존재해요. 이 때는 구현을 끝마치고서 테스트 코드를 작성하기도 합니다. 비록 TDD 과정이 아니기에 놓치는 부분들이 있겠지만, 적어도 TDD를 통해 얻고자 하는 이점들을 챙기고자 하고 있어요.

dye0p reviewed (commented):
2단계 리뷰 감사드립니다 :) 피드백 반영하여 수정 후 리뷰요청 하겠습니다!

dye0p committed:
refactor: absenceCount 재할당 로직 변경

- 기존의 파라미터로 받는 absenceCount가 재할당 되어서 예상하지 못한 변경이 있을 수 있었던 문제를 해결

dye0p committed:
test: 중복테스트 @Nested 어노테이션으로 한번에 묶음

dye0p committed:
refactor: 메서드 네이밍 명확하게 변경(findAttendanceBy -> findCrewAttendanceBy)

dye0p committed:
refactor: 캠퍼스 운영시간에 대한 검증 책임을 AttendanceDate로 이동

- 기존 캠퍼스 등교시간에 대한 책임을 Attendances에서 지고 있었지만 이건 적절하지 못한 책임이었음, 시간과 날짜에 관한 책임은 AttendanceDate에서 가지고 있는것이 더욱 적절한 책임이라 판단하여 관련 로직을 이동하였음

dye0p commented:
안녕하세요 이프! 이번이 마지막 리뷰요청이 될 것 같습니다! 😊

말씀해주신 피드백 최대한 반영해보았는데 적절하게 수정한건지 확신히 들지가 않은것 같아요..! 마음에 들지 않았던 부분은 Crew와 Attendance를 따로 분리하는게 더욱 좋았을 것 같다는 생각이 드는것 같습니다! 현재의 코드를 보면 Crew 객체가 아무일도 하지 않는 것 같은 느낌이 들었어요! 그래서 Attendances에 적절한 책임이동이 이루어지지 않은것 같습니다!

### 이번 미션을 통해서 느낀점
TDD 개발을 처음 해봤을때 어려움이 많이 느껴졌습니다! 특히 설계를 최소한만 가지고 가라는 의미를 잘 못 이해하였다고 생각이 들었고, 설계를 최소화 하라는 말은 객체에 대한 설계를 최소화 하고, 기능을 단위로 잘 분리하는 습관을 들여야 겠다고 느꼇어요! 

2단계 리뷰도 수고많으셨습니다 : )

sinb57 commented on src/test/java/domain/AttendanceStateTest.java:
```diff
@@ -3,33 +3,93 @@
+        assertThat(actual).isEqualTo(AttendanceState.LATE);
     }
 
+    @Nested
+    @DisplayName("출석 정보 계산 테스트")
+    class AttendanceStateFindTest {
```
`@Nested`를 사용하셨군요! 👍

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,147 @@
+            return;
+        }
+        LocalDateTime attendanceDateTime = inputAttendanceDateTime();
+        attendanceDate.validateInTime(attendanceDateTime.toLocalTime());
```
Controller에서 검증을 해야할까요? 출석 일시에 대한 값을 포장한 채 검증을 위임하길 바랬어요.

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,147 @@
+
+    private final InputView inputView;
+    private final OutputView outputView;
+    private final AttendanceDate attendanceDate;
+    private final Attendances attendances;
```
Controller가 해당 객체들을 인스턴스 필드로 지녀야 하는지 의문이에요

sinb57 commented on src/main/java/util/FileManager.java:
```diff
@@ -1,65 +1,103 @@
     private static final String SPLIT_DELIMITER = ",";
 
-    public static Attendance readFile(String filePath) {
+    public static Attendances readFile(final String filePath) {
```
FileManager의 메서드를 모두 static으로 만들어야 했을까요? 객체지향과 유틸리티, 그리고 static의 관계에 대해 알아보시면 좋겠어요~

sinb57 reviewed (approved):
안녕하세요 도기~ 2단계 진행하느라 고생 많으셨어요! 아쉽지만 리뷰 요청 마감일이 지나서 이만 머지할게요. 수고하셨습니다! 새롭게 시작하는 블랙잭 미션도 응원할게요~

---

> 마음에 들지 않았던 부분은 Crew와 Attendance를 따로 분리하는게 더욱 좋았을 것 같다는 생각이 드는것 같습니다! 현재의 코드를 보면 Crew 객체가 아무일도 하지 않는 것 같은 느낌이 들었어요! 그래서 Attendances에 적절한 책임이동이 이루어지지 않은것 같습니다!

Crew(크루)에 Attendances(개인 출석정보)를 매핑하여 Map<Crew, Attendances>를 만들어두었다면 Controller에 로직을 담지 않고, 적절한 책임을 지닌 객체를 만들어낼 수 있었을 거예요. 언젠가 기회가 된다면 출석 미션을 다시 해보길 바래요~ 그때 만약 리뷰가 필요하다면 DM 주세요~
